### DISCUSSION TITLE
Afterburner on Arduino NANO

### DISCUSSION THREAD

**Discussion by hubertushirsch** [ole00/afterburner#49] [Show and tell]
First of all, thank you very much for the great project.

I would like to show my afterburner here, which I built with an Arduino NANO V3. Since I have a lot of NANO clones left, but not a single UNO, I chose this construction.

- The NANO is on an I/O expansion board (€1.99), which is used as a NANO-to-UNO extender.
- Above it is a dummy board, which serves to stabilize the construction. It has no electrical function.
- The top level is the Afterburner itself.

The afterburner is not electrically modified, and the Arduino software is also original. The afterburner is equipped with through hole components, except the MOSFET. I chose a BSS138P because of the low RDS (0.9 ohms). It was only available as SMD.
A wire bridge had to be soldered on the I/O expansion board because 3V3 from the NANO socket was not routed to the appropriate UNO pin. That was the only modification I had to make.

![grafik](https://github.com/ole00/afterburner/assets/51236138/8d884d40-8174-4ba2-b913-d85acd6ce018)
The complete structure

![grafik](https://github.com/ole00/afterburner/assets/51236138/5b2d55af-cd79-4418-98f0-f835f0338ce8)
![grafik](https://github.com/ole00/afterburner/assets/51236138/f6379e3e-d101-4286-b358-c258ce29ba65)
The individual components

Regards
Hubert

**Comment by ole00** [comment]
Hi Hubert,
that's an interesting solution - thanks for sharing , it might help somebody to use their NANO board. MOSFET: yes, many mosfets would do the job, lower RDS is better of course. I used whatever I had available in my drawer. I'm glad the dual footprint option on the PCB is useful. Also, you soldered a pin header on VPP test holes - I left the holes empty and stick the multimeter probes directly to them when doing measurements.

I can see that the middle green board is there mainly because the NANO is not soldered directly on the bottom red board (so that you can unplug it if needed for other projects). Therefore the green board prevents NANO to collide with Afterburner's PCB. That makes sense. What you could possibly do is to push the mounting header pins on Afterburner a bit down, so that they are flush with the PCB  before soldering. It would give you ~ 1mm extra length of the pins. Or you could possibly use longer pin headers (like you used on the green board) and cut them to appropriate length. That way, you would not need to design the green board. Anyway - it seems your setup works fine for you, and that's the most important. 

 User @jalonsorv already mentioned (see #29) that (s)he redesigned Afterburner's PCB for NANO, but did not share the design (yet?).

**Comment by ole00** [comment]
Also, you use an alternative style of the R9 pot, that's good to know it fits in the footprint.
I forgot to ask: do you use external power (barrel jack) on your red board with Afterburner? Or do you power NANO via USB only?

**Comment by jalonsorv** [comment]
I do have the design, if anyone is interested, but it's based on the original? (simpler) design, also the way I wired it up only takes into account the GAL22V10 side of the diagram published, so I have no idea if it would work with any other GAL
![IMG_20231108_153448784](https://github.com/ole00/afterburner/assets/112061662/250855fd-d864-4f8d-9339-d3623d00075e)

**Comment by ole00** [comment]
@jalonsorv that looks neat and compact. You never know, somebody might be interested to use just ATF22V10. In theory ATF750C should work with that design once the full support is finished (or if the @nospam2000 Afterburner fork is used). Thanks for the info and the image.

**Comment by hubertushirsch** [comment]
> I forgot to ask: do you use external power (barrel jack) on your red board with Afterburner? Or do you power NANO via USB only?

Until now I have worked with an additional external power supply when reading or writing a GAL. I only did power via USB without a GAL chip to flash the software in the Arduino.
The design of @jalonsorv has no external power supply, it will probably work USB only for me too.

**Reply by jalonsorv** [reply]
I never considered that the USB power might not be enough, I just built it on a breadboard and once it worked I designed the PCB, maybe it does draw more than the 500mA that any USB port should supply, but USB ports on a computer usually have some headroom.

Maybe the new design that uses more components draws more current as well.

**Reply by hubertushirsch** [reply]
The 500mA from the PC or hub USB is only half the story.
In the NANO there is a Schottky diode between Vusb and Vcc (5V).
An SS1P3L is shown in the original circuit diagram (from arduino.cc). It has a forward current of at least If=1A.
Other diodes are installed in the NANO clones. You can find circuit diagrams with SD101CWS (If=20mA..30mA) and some with MBR0520LT1G (If=500mA).
For NANO clones with the SD101CWS diode, it is critical to power the NANO itself and the afterburner hardware via USB.
I just researched this on the internet and haven't yet checked how my NANO clones are equipped.

**EDIT:**
I just checked the NANO Clones in my stock. The diode from Vusb to Vcc (+5V) is marked "**SL**". That would have to be an SD1002 (20V/1A) Schottky diode from Taiwan Semiconductor (TSC).
![grafik](https://github.com/ole00/afterburner/assets/51236138/495bf020-c13a-4f12-a5ee-a315594b9cc7)
So I'll test something else without an external power supply.

**Comment by hubertushirsch** [comment]
> I can see that the middle green board is there mainly because the NANO is not soldered directly on the bottom red board (so that you can unplug it if needed for other projects).

The red board is bought ready. The connector for the NANO is already on there.
![grafik](https://github.com/ole00/afterburner/assets/51236138/00408091-c6d2-498c-8651-01ac0c3ee630)

**Comment by ole00** [comment]
@hubertushirsch I see, that explains there was no other option.

EDIT: one could possibly use something like this: 
https://www.pcbway.com/project/shareproject/Arduino_nano_to_uno_adaptor_board.html

**Reply by hubertushirsch** [reply]
Yes, thanks. I had already found this project. But having a circuit board manufactured in Germany or China for a single afterburner would have been several times more expensive.
Therefore I took a cheap ready-made module. Even if I don't need the pin headers for all the connections.
The NANO to UNO route was the decisive factor.

**Comment by hubertushirsch** [comment]
Last info:
Supplying with USB only delivers different (higher) voltages than with USB+external power supply.
The calibration only seems to be correct if the voltage supply is reproducible. I was not expecting that.
There may even be differences if I use different USB ports. PC, notebook, USB hub w/wo power supply.
I have to test it. Tomorrow.
```
** USB + ext. Power**
$ ./afterburner.exe -d COM15 m

VPP calib. offset: -5
VPP: 4.2 - 5.0V : 4.82
VPP: 9.0V : 8.98
VPP: 10.0V : 10.02
VPP: 12.0V : 11.96
VPP: 14.0V : 14.02
VPP: 16.0V : 15.90

**USB only**
$ ./afterburner.exe -d COM15 m

VPP calib. offset: -5
VPP: 4.2 - 5.0V : 4.47
VPP: 9.0V : 9.08
VPP: 10.0V : 10.13
VPP: 12.0V : 12.08
VPP: 14.0V : 14.17
VPP: 16.0V : 16.07
```

**Comment by hubertushirsch** [comment]
I would like to share some of the insights I have gained with my Afterburner/NANO construction.
In the following I want to call the NANO I/O-Expansion Shield (or NANO-2-UNO-Converter) **RedBoard**.

The external power connection of the **RedBoard** only supplies the **VIN** of the NANO.
The NANO's **5V** pin delivers 5V from the NANO's voltage regulator (LDO) back to the **RedBoard**.
If there is no external voltage on the **RedBoard** and therefore **VIN** is open, the **5V** pin carries approximately 4.7V when the NANO is powered via USB. (Loss due to UF of the Schottky diode).

On the **RedBoard** the **5V** pin is connected to the input of a 3.3V LDO. The generated 3.3V goes to the **3V3** pin of the NANO.
Since 3.3V is also generated on the NANO, either by the CH340G chip or the FT232RL chip (depending on the NANO model), two 3.3V outputs now work competitively against each other. Since I think this isn't good, I removed the 3.3V LDO from **RedBoard**.

See also the voltage rail diagram of NANO and **RedBoard**.
![grafik](https://github.com/ole00/afterburner/assets/51236138/edcbc895-e166-402b-b580-17f881f40aef)

Another change I made concerns the voltage measurement of the afterburner's VPP.
I think the generation of AREF from an LDO is too unstable in terms of temperature and dependence on the input voltage. The 3.3V output of a USB/serial converter was certainly never designed as a reference for a 10bit A/D converter.
That's why I set the reference voltage for the ADC to the internal 1.1V bandgap reference.
In the Arduino software I had to change line 308 in **aftb_vpp.h**
```analogReference(ANALOG_REF_EXTERNAL); //use 3V3 external reference```
to
```analogReference(INTERNAL); //use 1V1 internal reference```
R7 was removed from the Afterburner circuit board. The voltage divider R5/R6 was changed from 100kOhm/20kOhm to 20kOhm/1.3kOhm. The calibration was carried out without any problems.

**Note:** This change is only valid for the Arduino NANO and the Arduino UNO R3.
These two boards use the ATmega328P, which has an internal reference of 1.1V.
The UNO R4 uses a different MCU and has an internal reference voltage of 1.5V.

**Comment by ole00** [comment]
Thanks Hubert, that's interesting. It's good to verify the design and try different approaches.
 Just for the record - the GAL chips (at least those that I tested so far)  are fortunately quite tolerant to VPP. Until recently the code in Afterburner under-voltaged the desired VPP (programming voltage) by 2V during fuse writing. While this seemed to work OK on my Afterburner's board, it was on the very edge of fuse writing reliability. The latest code reduces the target VPP by 1V during fuse writing to improve the reliability. My point is, with the latest Afterburner's code the VPP calibration can be +/-0.1V off and the GALs should be programmed fine. Higher VPP accuracy does not seem to play a role in this project.

Unfortunately there is no public and official (from the GAL manufacturers) information about what the VPP should really be set for each type of GAL. The VPP value is not specified in any datasheets. We have a code from the other GAL projects that somehow calculates the required VPP, but I can't be certain such calculation formula is correct. What is known is that low VPP leads to fuse writing errors (verification fails) and high VPP may damage the GAL chip. In theory the GAL chips could be tested by series of runs, each run would use different VPP and based on the test results the VPP calculation formula could be verified. The problem is that the calculation of VPP (for non Atmel chips) is based on data stored in PES data (part of the fuse map) and that PES data can be overwritten to any value. So, if the PES data is (over)written by a wrong VPP calculation data, the VPP calculation will be incorrect.

**Reply by hubertushirsch** [reply]
Thanks for the reply. It's two things.
On the one hand, there is the determination of which VPP a specific GAL type requires.
Whether this can be taken from a data sheet, calculated from the PES or based on empirical values is irrelevant. At the end of the process there is a voltage value.
If the fuses of a PES were previously set incorrectly, the programmer's hardware cannot know this.
The problem may have to be solved in the software by manually forcing a VPP that is typical for this GAL type.
In the end, the programmer hardware must be able to reliably deliver the desired value.

On the other hand is the hardware design of the programmer.
The basis for reliable voltage provision is correct calibration, which requires reliable measurement.
I don't find deploying AREF from an LDO very good because there are a lot of dependencies.
In addition, the voltage division with R7 also needs to be questioned.
```
  3V3   ____    AREF   ____   
   o---|____|----o----|____|---
       R7 3k3         Ri 32k
```
Ri is specified in the data sheet with typ. 32kOhm, but there is no tolerance range and no temperature dependence specified in the data sheet. Even though R7 is a 1% resistor, deviations/fluctuations in Ri have a 10 times greater influence on AREF than deviations in R7.
This is my personal opinion and is by no means absolute wisdom.

I admit that my solution has a disadvantage. If an Arduino is used that has an MCU with a different internal reference (UNO R4),
the voltage divider R5/R6 must be dimensioned differently. I haven't calculated that for 1.5V internal reference yet.

I just wanted to introduce you to my implementation, as an alternative to the original circuit.
Since neither a change to the circuit board layout nor the software is necessary, I see this as an incentive for users to
to choose between the variants.
